File: backend/src/main/java/com/issuetracker/service/IssueService.java
package com.issuetracker.service;

import com.issuetracker.dto.CreateIssueRequest;
import com.issuetracker.dto.IssueDto;
import com.issuetracker.dto.IssueUpdateEvent;
import com.issuetracker.dto.PageResponse;
import com.issuetracker.model.Issue;
import com.issuetracker.model.IssuePriority;
import com.issuetracker.model.IssueStatus;
import com.issuetracker.model.Project;
import com.issuetracker.model.User;
import com.issuetracker.repository.IssueRepository;
import com.issuetracker.repository.ProjectRepository;
import com.issuetracker.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.messaging.simp.SimpMessagingTemplate;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.transaction.event.TransactionPhase;
import org.springframework.transaction.event.TransactionalEventListener;

import java.util.stream.Collectors;

@Service
public class IssueService {
    
    @Autowired
    private IssueRepository issueRepository;
    
    @Autowired
    private ProjectRepository projectRepository;
    
    @Autowired
    private UserRepository userRepository;
    
    @Autowired
    private SimpMessagingTemplate messagingTemplate;
    
    @Transactional
    public IssueDto createIssue(CreateIssueRequest request) {
        Project project = projectRepository.findById(request.getProjectId())
            .orElseThrow(() -> new RuntimeException("Project not found"));
        
        Issue issue = new Issue();
        issue.setTitle(request.getTitle());
        issue.setDescription(request.getDescription());
        issue.setStatus(request.getStatus() != null ? request.getStatus() : IssueStatus.OPEN);
        issue.setPriority(request.getPriority() != null ? request.getPriority() : IssuePriority.MEDIUM);
        issue.setProject(project);
        
        if (request.getAssigneeId() != null) {
            User assignee = userRepository.findById(request.getAssigneeId())
                .orElseThrow(() -> new RuntimeException("Assignee not found"));
            issue.setAssignee(assignee);
        }
        
        issue = issueRepository.save(issue);
        
        IssueDto dto = convertToDto(issue);
        
        // Publish WebSocket event after transaction commits
        IssueUpdateEvent event = new IssueUpdateEvent(
            "CREATED",
            dto.getId(),
            dto.getTitle(),
            dto.getStatus(),
            dto.getPriority(),
            dto.getProjectId()
        );
        messagingTemplate.convertAndSend("/topic/issues", event);
        
        return dto;
    }
    
    public PageResponse<IssueDto> getIssues(
        Integer page,
        Integer size,
        String sortBy,
        String sortDir,
        IssueStatus status,
        IssuePriority priority,
        Long assigneeId,
        Long projectId,
        String searchText
    ) {
        Sort sort = Sort.by(sortDir != null && sortDir.equalsIgnoreCase("desc") 
            ? Sort.Direction.DESC 
            : Sort.Direction.ASC, 
            sortBy != null ? sortBy : "createdAt");
        
        Pageable pageable = PageRequest.of(page != null ? page : 0, size != null ? size : 20, sort);
        
        Page<Issue> issuePage = issueRepository.findAllWithFilters(
            status,
            priority,
            assigneeId,
            projectId,
            searchText,
            pageable
        );
        
        return convertToPageResponse(issuePage);
    }
    
    public IssueDto getIssueById(Long id) {
        Issue issue = issueRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Issue not found"));
        
        return convertToDto(issue);
    }
    
    @Transactional
    public IssueDto updateIssue(Long id, CreateIssueRequest request) {
        Issue issue = issueRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Issue not found"));
        
        issue.setTitle(request.getTitle());
        issue.setDescription(request.getDescription());
        issue.setStatus(request.getStatus());
        issue.setPriority(request.getPriority());
        
        if (request.getAssigneeId() != null) {
            User assignee = userRepository.findById(request.getAssigneeId())
                .orElseThrow(() -> new RuntimeException("Assignee not found"));
            issue.setAssignee(assignee);
        } else {
            issue.setAssignee(null);
        }
        
        issue = issueRepository.save(issue);
        
        IssueDto dto = convertToDto(issue);
        
        // Publish WebSocket event
        IssueUpdateEvent event = new IssueUpdateEvent(
            "UPDATED",
            dto.getId(),
            dto.getTitle(),
            dto.getStatus(),
            dto.getPriority(),
            dto.getProjectId()
        );
        messagingTemplate.convertAndSend("/topic/issues", event);
        
        return dto;
    }
    
    @Transactional
    public void deleteIssue(Long id) {
        Issue issue = issueRepository.findById(id)
            .orElseThrow(() -> new RuntimeException("Issue not found"));
        
        Long projectId = issue.getProject().getId();
        String title = issue.getTitle();
        
        issueRepository.delete(issue);
        
        // Publish WebSocket event
        IssueUpdateEvent event = new IssueUpdateEvent(
            "DELETED",
            id,
            title,
            issue.getStatus(),
            issue.getPriority(),
            projectId
        );
        messagingTemplate.convertAndSend("/topic/issues", event);
    }
    
    private IssueDto convertToDto(Issue issue) {
        IssueDto dto = new IssueDto();
        dto.setId(issue.getId());
        dto.setTitle(issue.getTitle());
        dto.setDescription(issue.getDescription());
        dto.setStatus(issue.getStatus());
        dto.setPriority(issue.getPriority());
        dto.setProjectId(issue.getProject().getId());
        dto.setProjectName(issue.getProject().getName());
        
        if (issue.getAssignee() != null) {
            dto.setAssigneeId(issue.getAssignee().getId());
            dto.setAssigneeName(issue.getAssignee().getName());
        }
        
        dto.setCreatedAt(issue.getCreatedAt());
        dto.setUpdatedAt(issue.getUpdatedAt());
        
        return dto;
    }
    
    private PageResponse<IssueDto> convertToPageResponse(Page<Issue> issuePage) {
        return new PageResponse<>(
            issuePage.getContent().stream()
                .map(this::convertToDto)
                .collect(Collectors.toList()),
            issuePage.getNumber(),
            issuePage.getSize(),
            issuePage.getTotalElements(),
            issuePage.getTotalPages(),
            issuePage.isLast()
        );
    }
}

File: frontend/src/app/services/issue.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';

export interface Issue {
  id: number;
  title: string;
  description: string;
  status: 'OPEN' | 'IN_PROGRESS' | 'CLOSED';
  priority: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  projectId: number;
  projectName: string;
  assigneeId?: number;
  assigneeName?: string;
  createdAt: string;
  updatedAt: string;
}

export interface CreateIssueRequest {
  title: string;
  description?: string;
  status?: 'OPEN' | 'IN_PROGRESS' | 'CLOSED';
  priority?: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  projectId: number;
  assigneeId?: number;
}

export interface PageResponse<T> {
  content: T[];
  page: number;
  size: number;
  totalElements: number;
  totalPages: number;
  last: boolean;
}

export interface IssueFilters {
  page?: number;
  size?: number;
  sortBy?: string;
  sortDir?: 'asc' | 'desc';
  status?: 'OPEN' | 'IN_PROGRESS' | 'CLOSED';
  priority?: 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL';
  assigneeId?: number;
  projectId?: number;
  searchText?: string;
}

@Injectable({
  providedIn: 'root'
})
export class IssueService {
  private apiUrl = 'http://localhost:8080/api/issues';

  constructor(private http: HttpClient) {}

  getIssues(filters: IssueFilters = {}): Observable<PageResponse<Issue>> {
    let params = new HttpParams();
    
    if (filters.page !== undefined) params = params.set('page', filters.page.toString());
    if (filters.size !== undefined) params = params.set('size', filters.size.toString());
    if (filters.sortBy) params = params.set('sortBy', filters.sortBy);
    if (filters.sortDir) params = params.set('sortDir', filters.sortDir);
    if (filters.status) params = params.set('status', filters.status);
    if (filters.priority) params = params.set('priority', filters.priority);
    if (filters.assigneeId !== undefined) params = params.set('assigneeId', filters.assigneeId.toString());
    if (filters.projectId !== undefined) params = params.set('projectId', filters.projectId.toString());
    if (filters.searchText) params = params.set('searchText', filters.searchText);

    return this.http.get<PageResponse<Issue>>(this.apiUrl, { params });
  }

  getIssueById(id: number): Observable<Issue> {
    return this.http.get<Issue>(`${this.apiUrl}/${id}`);
  }

  createIssue(issue: CreateIssueRequest): Observable<Issue> {
    return this.http.post<Issue>(this.apiUrl, issue);
  }

  updateIssue(id: number, issue: CreateIssueRequest): Observable<Issue> {
    return this.http.put<Issue>(`${this.apiUrl}/${id}`, issue);
  }

  deleteIssue(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
